File: frontend/src/services/subscriptions.js
import axios from 'axios';

const API_KEY = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InliZGV4ZHVma25kZGJjaGV3ZGttIiwicm9sZSI6ImFub24iLCJpYXQiOjE3MjIyNDcyODUsImV4cCI6MjAzNzgyMzI4NX0.k62P6W0dO7E_BJpy9zVfm-HYhVlE3vrRldMo_1uUa9U';

export const fetchSubscriptions = () => {
    return api.get('/subscriptions', {
        headers: {
            'Authorization': `Bearer ${API_KEY}`
        }
    });
};

export const createSubscription = (data) => {
    return api.post('/subscriptions', data, {
        headers: {
            'Authorization': `Bearer ${API_KEY}`
        }
    });
};

File: fastapi-react-app/routers/subscriptions.py
from fastapi import APIRouter, HTTPException, Depends
import httpx
from models.schemas import Subscription
from config import API_URL, API_KEY
from pydantic import BaseModel
from dependencies import verify_token

router = APIRouter()

class SubscriptionCreate(BaseModel):
    id: int
    preference_id: int
    enabled: bool
    user_id: str

class SubscriptionUpdate(BaseModel):
    id: int
    preference_id: int
    enabled: bool
    user_id: str

async def get_client():
    async with httpx.AsyncClient() as client:
        yield client

@router.get("/", response_model=list[Subscription])
async def get_subscriptions(client: httpx.AsyncClient = Depends(get_client), token: str = Depends(verify_token)):
    headers = {
        "apikey": token,
        "Authorization": f"Bearer {token}"
    }
    response = await client.get(f"{API_URL}/subscriptions?select=*", headers=headers)
    if response.status_code != 200:
        raise HTTPException(status_code=response.status_code, detail=response.text)
    return response.json()

@router.post("/", response_model=None)
async def insert_subscription(subscription: SubscriptionCreate, client: httpx.AsyncClient = Depends(get_client), token: str = Depends(verify_token)):
    headers = {
        "apikey": token,
        "Authorization": f"Bearer {token}",
        "Content-Type": "application/json",
        "Prefer": "return=minimal"
    }
    response = await client.post(
        f"{API_URL}/subscriptions",
        headers=headers,
        json=subscription.dict()
    )
    if response.status_code != 201:
        raise HTTPException(status_code=response.status_code, detail=response.text)
    
@router.delete("/{id_column_value}", response_model=None)
async def delete_subscription(id_column_value: int, client: httpx.AsyncClient = Depends(get_client), token: str = Depends(verify_token)):
    headers = {
        "apikey": token,
        "Authorization": f"Bearer {token}"
    }
    response = await client.delete(
        f"{API_URL}/subscriptions?id=eq.{id_column_value}",
        headers=headers
    )
    if response.status_code != 204:
        raise HTTPException(status_code=response.status_code, detail=response.text)
    
@router.patch("/{id_column_value}", response_model=None)
async def update_preference(id_column_value: str, preference: SubscriptionUpdate, client: httpx.AsyncClient = Depends(get_client), token: str = Depends(verify_token)):
    headers = {
        "apikey": token,
        "Authorization": f"Bearer {token}",
        "Content-Type": "application/json",
        "Prefer": "return=minimal"
    }
    response = await client.patch(
        f"{API_URL}/subscriptions?id=eq.{id_column_value}",
        headers=headers,
        json=preference.dict()
    )
    if response.status_code != 204:
        raise HTTPException(status_code=response.status_code, detail=response.text)